gdal/frmts/wcs/wcsdataset201.cpp

```diff
@@ -0,0 +1,781 @@
+    extent.push_back(adfGeoTransform[3] +
+                     (nYOff) * adfGeoTransform[5]);
+
+    extent[2] -= adfGeoTransform[1] * 0.5;
+    extent[0] += adfGeoTransform[1] * 0.5;
+    extent[1] -= adfGeoTransform[5] * 0.5;
+    extent[3] += adfGeoTransform[5] * 0.5;
+
+    return extent;
+}
+
+/************************************************************************/
+/*                        GetCoverageRequest()                          */
+/*                                                                      */
+/************************************************************************/
+
+CPLString WCSDataset201::GetCoverageRequest(bool scaled,
+                                            int nBufXSize, int nBufYSize,
+                                            std::vector<double> extent,
+                                            CPL_UNUSED CPLString osBandList)
+{
+    CPLString request = CPLGetXMLValue(psService, "ServiceURL", "");
+    request = CPLURLAddKVP(request, "SERVICE", "WCS");
+    request += "&REQUEST=GetCoverage";
+    request += "&VERSION=" + CPLString(CPLGetXMLValue(psService, "Version", ""));
+    request += "&COVERAGEID=" + URLEncode(CPLGetXMLValue(psService, "CoverageName", ""));
+    if (!native_crs) {
+        CPLString crs = URLEncode(CPLGetXMLValue(psService, "SRS", ""));
+        request += "&OUTPUTCRS=" + crs;
+        request += "&SUBSETTINGCRS=" + crs;
+    }
+    request += "&FORMAT=" + URLEncode(CPLGetXMLValue(psService, "PreferredFormat", ""));
+
+    // todo: updatesequence
+
+    std::vector<CPLString> domain = Split(CPLGetXMLValue(psService, "Domain", ""), ",");
+    const char *x = domain[0].c_str();
+    const char *y = domain[1].c_str();
+    if (CPLGetXMLBoolean(psService, "SubsetAxisSwap")) {
+        const char *tmp = x;
+        x = y;
+        y = tmp;
+    }
+
+    CPLString a = CPLString().Printf(
+        "%.17g", MAX(adfGeoTransform[0], extent[0]));
+    CPLString b = CPLString().Printf(
+        "%.17g", MIN(adfGeoTransform[0] + nRasterXSize * adfGeoTransform[1], extent[2]));
+
+    request += CPLString().Printf("&SUBSET=%s%%28%s,%s%%29", x, a.c_str(), b.c_str());
+
+    a = CPLString().Printf(
+        "%.17g", MAX(adfGeoTransform[3] + nRasterYSize * adfGeoTransform[5], extent[1]));
+    b = CPLString().Printf(
+        "%.17g", MIN(adfGeoTransform[3], extent[3]));
+
+    request += CPLString().Printf("&SUBSET=%s%%28%s,%s%%29", y, a.c_str(), b.c_str());
+
+    // set subsets for axis other than x/y
+    std::vector<CPLString> dimensions = Split(CPLGetXMLValue(psService, "Dimensions", ""), ";");
+    for (unsigned int i = 0; i < dimensions.size(); ++i) {
+        size_t pos = dimensions[i].find("(");
+        CPLString dim = dimensions[i].substr(0, pos);
+        if (IndexOf(dim, domain) != -1) {
+            continue;
+        }
+        std::vector<CPLString> params = Split(FromParenthesis(dimensions[i]), ",");
+        request += "&SUBSET=" + dim + "%28";
```

Comment:
you could probably use CPLEscapeString(str, -1, CPLES_URL) to do the URL escaping